File: 0394-decode-string/0394-decode-string.java
class Solution {
    public String decodeString(String s) {
        if(s == null || s.length() == 0){
            return "";
        }

        Stack<Integer> numSt = new Stack<>();
        Stack<String> strSt = new Stack<>();
        String curr = "";
        int num = 0;

        for(int i = 0; i < s.length(); i++){
            char ch = s.charAt(i);

            if(Character.isDigit(ch)){
                num = num * 10 + ch - '0';
            }
            else if (ch == '['){
                numSt.add(num);
                strSt.add(curr);
                curr = "";
                num = 0;
            }
            else if(ch == ']'){
                int times = numSt.pop();
                StringBuilder newStr = new StringBuilder();
                for(int time = 0; time < times; time++){
                    newStr.append(curr);
                }
                curr = strSt.pop();
                curr += newStr;
            }
            else{
                curr += ch;
            }
        }

        return curr;

    }
}

File: 0394-decode-string/0394-decode-string.py
class Solution:
    def decodeString(self, s: str) -> str:
        if len(s) == 0:
            return ""

        # numSt = []
        # charSt = []
        nums = 0
        result = [""]
        
        for char in s:

            if char.isdigit():
                nums = nums * 10 + (int(char))

            elif char == "[":
                result.append(nums)
                result.append("")
                nums = 0

            elif char == "]":
                # print(numSt,charSt,i)
                strSt = result.pop()
                times = result.pop()
            
                result[-1] += strSt * times
            else:
                result[-1] += char
        # print(result)
        return "".join(result)

